// @alteryx/ui/ListBox/ListBoxFooter.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _Typography = _interopRequireDefault(require("../Typography"));

var _Divider = _interopRequireDefault(require("../Divider"));

var _makeStyles = _interopRequireDefault(require("@material-ui/core/styles/makeStyles"));

var _reactIntl = require("react-intl");

var useStyles = (0, _makeStyles.default)(function (_ref) {
  var spacing = _ref.spacing;
  return {
    message: {
      paddingTop: spacing(2)
    }
  };
});

var _ref3 = /*#__PURE__*/_react.default.createElement(_Divider.default, {
  light: true
});

function ListBoxFooter(_ref2) {
  var checkedCount = _ref2.checkedCount,
      classes = _ref2.classes,
      footerText = _ref2.footerText,
      totalCount = _ref2.totalCount;
  var textClass = useStyles();
  var intl = (0, _reactIntl.useIntl)();

  var getFooterText = function getFooterText() {
    var messages = (0, _reactIntl.defineMessages)({
      footer: {
        id: 'footer.text',
        defaultMessage: footerText
      }
    });
    var total = intl.formatNumber(totalCount);
    var selected = intl.formatNumber(checkedCount);
    var intlFormatMessage = intl.formatMessage(messages.footer, {
      checked: /*#__PURE__*/_react.default.createElement("b", null, selected),
      total: /*#__PURE__*/_react.default.createElement("b", null, total)
    });
    var messageToProcess = /*#__PURE__*/_react.default.isValidElement(intlFormatMessage) && Array.isArray(intlFormatMessage.props.children) ? intlFormatMessage.props.children : intlFormatMessage;
    var stringMessage = Array.isArray(messageToProcess) ? messageToProcess.map(function (child) {
      return /*#__PURE__*/_react.default.isValidElement(child) ? child.props.children : child;
    }).join('') : messageToProcess;
    return stringMessage;
  };

  return /*#__PURE__*/_react.default.createElement("div", {
    className: classes.footer
  }, _ref3, /*#__PURE__*/_react.default.createElement(_Typography.default, {
    className: textClass.message,
    variant: "subtitle2"
  }, getFooterText()));
}

ListBoxFooter.defaultProps = {
  footerText: '{checked} of {total} Checked'
};
ListBoxFooter.propTypes = process.env.NODE_ENV !== "production" ? {
  checkedCount: _propTypes.default.number.isRequired,
  footerText: _propTypes.default.string,
  totalCount: _propTypes.default.number.isRequired
} : {};
var _default = ListBoxFooter;
exports.default = _default;
