// @alteryx/ui/ListBox/ListBoxSearch.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _icons = require("@alteryx/icons");

var _FormControl = _interopRequireDefault(require("../FormControl"));

var _FilledInput = _interopRequireDefault(require("../FilledInput"));

var _InputAdornment = _interopRequireDefault(require("../InputAdornment"));

var _useTheme2 = _interopRequireDefault(require("@material-ui/core/styles/useTheme"));

var propTypes = process.env.NODE_ENV !== "production" ? {
  onChange: _propTypes.default.func,
  searchPlaceholderText: _propTypes.default.string.isRequired,
  title: _propTypes.default.node
} : {};
var defaultProps = {
  onChange: function onChange() {},
  title: undefined
};

function ListBoxSearch(_ref) {
  var classes = _ref.classes,
      searchPlaceholderText = _ref.searchPlaceholderText,
      _onChange = _ref.onChange,
      title = _ref.title;

  var _useTheme = (0, _useTheme2.default)(),
      palette = _useTheme.palette;

  return /*#__PURE__*/_react.default.createElement(_FormControl.default, {
    classes: {
      root: classes.searchFormControl
    }
  }, /*#__PURE__*/_react.default.createElement(_FilledInput.default, {
    classes: {
      root: classes.searchInput
    },
    inputProps: {
      'aria-labelledby': title
    },
    onChange: function onChange(_ref2) {
      var keyword = _ref2.target.value;
      return _onChange(keyword);
    },
    placeholder: searchPlaceholderText,
    role: "search",
    startAdornment: /*#__PURE__*/_react.default.createElement(_InputAdornment.default, {
      position: "start"
    }, /*#__PURE__*/_react.default.createElement(_icons.Search, {
      color: palette.type === 'dark' ? palette.text.secondary : palette.blueGrey[500],
      size: "small"
    }))
  }));
}

ListBoxSearch.propTypes = process.env.NODE_ENV !== "production" ? propTypes : {};
ListBoxSearch.defaultProps = defaultProps;
var _default = ListBoxSearch;
exports.default = _default;
