// @alteryx/ui/NumericInput/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.INTERMEDIATE_FORMATS = exports.FORMAT_POSITIVE = exports.FORMAT_FLOAT = exports.FORMAT_INTEGER = exports.SPINNER_MOUSE_INTERVAL = exports.SPINNER_KEYBOARD_INTERVAL = exports.SPINNER_CHANGE_DELAY = exports.CONTROL_KEYS = exports.REMOVAL_KEYS = void 0;

var _keycodeJs = require("keycode-js");

var REMOVAL_KEYS = [_keycodeJs.KEY_BACK_SPACE, _keycodeJs.KEY_DELETE];
exports.REMOVAL_KEYS = REMOVAL_KEYS;
var CONTROL_KEYS = [_keycodeJs.KEY_UP, _keycodeJs.KEY_DOWN];
exports.CONTROL_KEYS = CONTROL_KEYS;
var SPINNER_CHANGE_DELAY = 300;
exports.SPINNER_CHANGE_DELAY = SPINNER_CHANGE_DELAY;
var SPINNER_KEYBOARD_INTERVAL = 100;
exports.SPINNER_KEYBOARD_INTERVAL = SPINNER_KEYBOARD_INTERVAL;
var SPINNER_MOUSE_INTERVAL = 150;
exports.SPINNER_MOUSE_INTERVAL = SPINNER_MOUSE_INTERVAL;
var FORMAT_INTEGER = /^-{0,1}[1-9][0-9]*$/;
exports.FORMAT_INTEGER = FORMAT_INTEGER;
var FORMAT_FLOAT = /(^-{0,1}[1-9][0-9]*[.]{0,1}[0-9]*$)|(^-{0,1}[0-9]([.][0-9]*)?$)/;
exports.FORMAT_FLOAT = FORMAT_FLOAT;
var FORMAT_POSITIVE = /^[^-]*$/;
exports.FORMAT_POSITIVE = FORMAT_POSITIVE;
var INTERMEDIATE_FORMATS = ['', '-', '0'];
exports.INTERMEDIATE_FORMATS = INTERMEDIATE_FORMATS;
