// @alteryx/ui/NumericInput/NumericSpinner.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _lodash = require("lodash");

var _Button = _interopRequireDefault(require("../Button"));

var _makeStyles = _interopRequireDefault(require("@material-ui/core/styles/makeStyles"));

var _createStyles = _interopRequireDefault(require("@material-ui/core/styles/createStyles"));

var _icons = require("@alteryx/icons");

var _numericHoc = _interopRequireDefault(require("./numericHoc"));

var propTypes = process.env.NODE_ENV !== "production" ? {
  decrementButtonText: _propTypes.default.string,
  disabled: _propTypes.default.bool,
  endAdornment: _propTypes.default.element,
  error: _propTypes.default.bool,
  focused: _propTypes.default.bool,
  incrementButtonText: _propTypes.default.string,
  numericStart: _propTypes.default.func,
  numericStop: _propTypes.default.func,
  readOnly: _propTypes.default.bool
} : {};
var defaultProps = {
  disabled: undefined,
  endAdornment: null,
  error: undefined,
  focused: undefined,
  numericStart: _lodash.noop,
  numericStop: _lodash.noop,
  readOnly: undefined
};
var useStyles = (0, _makeStyles.default)(function () {
  return (0, _createStyles.default)({
    numericSpinnerRoot: {
      paddingRight: 2
    }
  });
});

function NumericSpinner(_ref) {
  var classes = _ref.classes,
      decrementButtonText = _ref.decrementButtonText,
      disabled = _ref.disabled,
      endAdornment = _ref.endAdornment,
      incrementButtonText = _ref.incrementButtonText,
      numericStart = _ref.numericStart,
      numericStop = _ref.numericStop,
      readOnly = _ref.readOnly;
  var spinnerClasses = useStyles();
  var incrementButton = classes.incrementButton,
      decrementButton = classes.decrementButton;
  var buttonVariant = disabled ? 'text' : 'contained';
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, endAdornment, /*#__PURE__*/_react.default.createElement("div", {
    className: spinnerClasses.numericSpinnerRoot
  }, /*#__PURE__*/_react.default.createElement(_Button.default, {
    "aria-label": incrementButtonText,
    className: incrementButton,
    disabled: disabled || readOnly,
    onMouseDown: numericStart(true),
    onMouseLeave: numericStop,
    onMouseUp: numericStop,
    tabIndex: -1,
    variant: buttonVariant
  }, /*#__PURE__*/_react.default.createElement(_icons.ChevronUp, {
    className: classes.incrementIcon,
    size: "small"
  })), /*#__PURE__*/_react.default.createElement(_Button.default, {
    "aria-label": decrementButtonText,
    className: decrementButton,
    disabled: disabled || readOnly,
    onMouseDown: numericStart(false),
    onMouseLeave: numericStop,
    onMouseUp: numericStop,
    tabIndex: -1,
    variant: buttonVariant
  }, /*#__PURE__*/_react.default.createElement(_icons.ChevronDown, {
    className: classes.decrementIcon,
    size: "small"
  }))));
}

NumericSpinner.propTypes = process.env.NODE_ENV !== "production" ? propTypes : {};
NumericSpinner.defaultProps = defaultProps;

var _default = (0, _numericHoc.default)(NumericSpinner);

exports.default = _default;
