// @alteryx/ui/theme/overrides/alert.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _styles = require("../../styles");

/* eslint-disable complexity */
var MuiAlert = function MuiAlert(_ref) {
  var palette = _ref.palette,
      spacing = _ref.spacing,
      typography = _ref.typography;
  var standardOpacity = palette.type === 'dark' ? 0.15 : 0.07;
  var outlinedOpacity = palette.type === 'dark' ? 0.2 : 0.07;
  var filledOpacity = palette.type === 'dark' ? 0.08 : 0.12;
  var focusShadow = palette.getFocusedShadowStyle();
  return {
    root: {
      borderRadius: spacing(1),
      fontSize: '.8125rem',
      lineHeight: 1.54,
      padding: spacing(0.5, 2, 0.5, 2.5)
    },
    action: {
      marginRight: 0,
      '& button': {
        paddingLeft: spacing(1),
        paddingRight: spacing(1),
        '& .MuiSvgIcon-root': {
          fontSize: typography.pxToRem(16)
        },
        '&:focus': {
          boxShadow: focusShadow
        }
      },
      '& .MuiDivider-vertical': {
        marginLeft: spacing(2),
        marginRight: spacing(2)
      },
      '& .MuiSvgIcon-root': {
        padding: 1
      }
    },
    icon: {
      padding: spacing(2.75, 0),
      marginRight: spacing(1.5)
    },
    standardInfo: {
      color: palette.type === 'dark' ? palette.info.light : palette.info.dark,
      backgroundColor: palette.type === 'dark' ? '#1a2a43' : '#e3f0fc',
      '& .MuiDivider-vertical': {
        backgroundColor: (0, _styles.fade)(palette.info.main, standardOpacity)
      },
      '& $icon': {
        color: palette.type === 'dark' ? palette.info.light : palette.info.dark
      }
    },
    standardError: {
      color: palette.type === 'dark' ? palette.error.light : palette.error.dark,
      backgroundColor: palette.type === 'dark' ? '#3e222a' : '#fbebed',
      '& .MuiDivider-vertical': {
        backgroundColor: (0, _styles.fade)(palette.error.main, standardOpacity)
      },
      '& $icon': {
        color: palette.type === 'dark' ? palette.error.light : palette.error.dark
      }
    },
    standardWarning: {
      color: palette.type === 'dark' ? palette.warning.light : palette.warning.dark,
      backgroundColor: palette.type === 'dark' ? '#3e2724' : '#fcede9',
      '& .MuiDivider-vertical': {
        backgroundColor: (0, _styles.fade)(palette.warning.main, standardOpacity)
      },
      '& $icon': {
        color: palette.type === 'dark' ? palette.warning.light : palette.warning.dark
      }
    },
    standardSuccess: {
      color: palette.type === 'dark' ? palette.success.light : palette.success.dark,
      backgroundColor: palette.type === 'dark' ? '#193230' : '#e0f4ef',
      '& .MuiDivider-vertical': {
        backgroundColor: (0, _styles.fade)(palette.success.main, standardOpacity)
      },
      '& $icon': {
        color: palette.type === 'dark' ? palette.success.light : palette.success.dark
      }
    },
    outlinedInfo: {
      padding: spacing(0.25, 2, 0.25, 2.5),
      color: palette.type === 'dark' ? palette.info.light : palette.info.dark,
      borderStyle: 'solid',
      borderWidth: 1,
      borderColor: palette.type === 'dark' ? (0, _styles.fade)(palette.info.main, 0.4) : (0, _styles.fade)(palette.info.main, 0.15),
      '& .MuiDivider-vertical': {
        backgroundColor: (0, _styles.fade)(palette.info.main, outlinedOpacity)
      },
      '& $icon': {
        color: palette.type === 'dark' ? palette.info.light : palette.info.dark
      }
    },
    outlinedError: {
      padding: spacing(0.25, 2, 0.25, 2.5),
      color: palette.type === 'dark' ? palette.error.light : palette.error.dark,
      borderStyle: 'solid',
      borderWidth: 1,
      borderColor: palette.type === 'dark' ? (0, _styles.fade)(palette.error.main, 0.4) : (0, _styles.fade)(palette.error.main, 0.15),
      '& .MuiDivider-vertical': {
        backgroundColor: (0, _styles.fade)(palette.error.main, outlinedOpacity)
      },
      '& $icon': {
        color: palette.type === 'dark' ? palette.error.light : palette.error.dark
      }
    },
    outlinedWarning: {
      padding: spacing(0.25, 2, 0.25, 2.5),
      color: palette.type === 'dark' ? palette.warning.light : palette.warning.dark,
      borderStyle: 'solid',
      borderWidth: 1,
      borderColor: palette.type === 'dark' ? (0, _styles.fade)(palette.warning.main, 0.4) : (0, _styles.fade)(palette.warning.main, 0.15),
      '& .MuiDivider-vertical': {
        backgroundColor: (0, _styles.fade)(palette.warning.main, outlinedOpacity)
      },
      '& $icon': {
        color: palette.type === 'dark' ? palette.warning.light : palette.warning.dark
      }
    },
    outlinedSuccess: {
      padding: spacing(0.25, 2, 0.25, 2.5),
      color: palette.type === 'dark' ? palette.success.light : palette.success.dark,
      borderStyle: 'solid',
      borderWidth: 1,
      borderColor: palette.type === 'dark' ? (0, _styles.fade)(palette.success.main, 0.4) : (0, _styles.fade)(palette.success.main, 0.15),
      '& .MuiDivider-vertical': {
        backgroundColor: (0, _styles.fade)(palette.success.main, outlinedOpacity)
      },
      '& $icon': {
        color: palette.type === 'dark' ? palette.success.light : palette.success.dark
      }
    },
    filledInfo: {
      backgroundColor: palette.info.main,
      borderRadius: spacing(1.5),
      '& .MuiDivider-vertical': {
        backgroundColor: (0, _styles.fade)(palette.common.white, filledOpacity)
      },
      '& $icon': {
        opacity: 0.7
      }
    },
    filledError: {
      backgroundColor: palette.error.main,
      borderRadius: spacing(1.5),
      '& .MuiDivider-vertical': {
        backgroundColor: (0, _styles.fade)(palette.common.white, filledOpacity)
      },
      '& $icon': {
        opacity: 0.7
      }
    },
    filledWarning: {
      backgroundColor: palette.warning.main,
      borderRadius: spacing(1.5),
      '& .MuiDivider-vertical': {
        backgroundColor: (0, _styles.fade)(palette.common.white, filledOpacity)
      },
      '& $icon': {
        opacity: 0.7
      }
    },
    filledSuccess: {
      backgroundColor: palette.success.main,
      borderRadius: spacing(1.5),
      '& .MuiDivider-vertical': {
        backgroundColor: (0, _styles.fade)(palette.common.white, filledOpacity)
      },
      '& $icon': {
        opacity: 0.7
      }
    }
  };
};

var MuiAlertTitle = function MuiAlertTitle() {
  return {
    root: {
      marginTop: 0,
      fontWeight: 600
    }
  };
};

var _default = function _default(theme) {
  return {
    MuiAlert: MuiAlert(theme),
    MuiAlertTitle: MuiAlertTitle()
  };
};

exports.default = _default;
