// @alteryx/ui/theme/overrides/dateTimePicker.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var MPickersToolbar = function MPickersToolbar(theme) {
  return {
    toolbar: {
      height: undefined,
      minHeight: 0,
      padding: theme.spacing(4)
    }
  };
};

var MPickersTimePickerToolbar = function MPickersTimePickerToolbar() {
  return {
    separator: {
      marginTop: 'auto',
      marginBottom: 'auto'
    }
  };
};

var MPickerDTToolbar = function MPickerDTToolbar(theme) {
  return {
    separator: {
      margin: theme.spacing(0, 1, 1.75, 0.5)
    }
  };
};

var MPickersClockNumber = function MPickersClockNumber() {
  return {
    clockNumber: {
      height: '30px'
    }
  };
};

var MPickersClockPointer = function MPickersClockPointer(theme) {
  return {
    thumb: {
      backgroundColor: theme.palette.primary.main
    }
  };
};

var MPickerDTTabs = function MPickerDTTabs(theme) {
  return {
    tabs: {
      color: theme.palette.getContrastText(theme.palette.background.paper),
      backgroundColor: theme.palette.background.paper
    }
  };
};

var MPickersDay = function MPickersDay(theme) {
  return {
    current: {
      border: "1px solid ".concat(theme.palette.blueGrey[200]),
      color: theme.palette.text.primary
    },
    day: {
      borderRadius: '50%',
      '&:hover': {
        color: theme.palette.text.primary,
        backgroundColor: theme.palette.action.hover
      }
    },
    daySelected: {
      backgroundColor: theme.palette.action.selected,
      color: theme.palette.text.primary,
      '&:hover': {
        backgroundColor: theme.palette.action.selected,
        color: theme.palette.text.primary
      },
      '&:focus': {
        backgroundColor: theme.palette.action.selected,
        color: theme.palette.text.primary
      }
    }
  };
};

var _default = function _default(theme) {
  return {
    MuiPickersToolbar: MPickersToolbar(theme),
    MuiPickersTimePickerToolbar: MPickersTimePickerToolbar(),
    MuiPickerDTToolbar: MPickerDTToolbar(theme),
    MuiPickersClockNumber: MPickersClockNumber(),
    MuiPickersClockPointer: MPickersClockPointer(theme),
    MuiPickerDTTabs: MPickerDTTabs(theme),
    MuiPickersDay: MPickersDay(theme)
  };
};

exports.default = _default;
