// @alteryx/ui/theme/overrides/timeline.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var MTimelineConnector = function MTimelineConnector(_ref) {
  var palette = _ref.palette;
  return {
    root: {
      width: 1,
      backgroundColor: palette.divider
    }
  };
};

var MTimelineDot = function MTimelineDot(_ref2) {
  var palette = _ref2.palette;
  return {
    root: {
      boxShadow: 'none'
    },
    defaultGrey: {
      backgroundColor: palette.blueGrey[500]
    }
  };
};

var MTimelineContent = function MTimelineContent(_ref3) {
  var spacing = _ref3.spacing;
  return {
    root: {
      padding: spacing(1, 3, 1.5, 3)
    }
  };
};

var MTimelineItem = function MTimelineItem(_ref4) {
  var spacing = _ref4.spacing;
  return {
    missingOppositeContent: {
      '&::before': {
        padding: spacing(1.5, 3)
      }
    }
  };
};

var MTimelineOppositeContent = function MTimelineOppositeContent(_ref5) {
  var spacing = _ref5.spacing;
  return {
    root: {
      padding: spacing(1.5, 3)
    }
  };
};

var _default = function _default(theme) {
  return {
    MuiTimelineConnector: MTimelineConnector(theme),
    MuiTimelineDot: MTimelineDot(theme),
    MuiTimelineContent: MTimelineContent(theme),
    MuiTimelineItem: MTimelineItem(theme),
    MuiTimelineOppositeContent: MTimelineOppositeContent(theme)
  };
};

exports.default = _default;
