// @alteryx/ui/TextField/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _TextField = _interopRequireDefault(require("@material-ui/core/TextField"));

var _utils = require("../utils");

var _styles = require("../styles");

/* eslint-disable no-console */
var useStyles = (0, _styles.makeStyles)({
  root: {
    borderColor: 'transparent',
    backgroundColor: 'transparent',
    '&:hover': {
      borderColor: 'transparent'
    }
  }
});

var TextField = /*#__PURE__*/_react.default.forwardRef(function (props, ref) {
  var variant = props.variant,
      InputProps = props.InputProps,
      rest = (0, _objectWithoutProperties2.default)(props, ["variant", "InputProps"]);

  if (variant === 'borderless') {
    var borderlessClasses = useStyles();
    var InputPropsClasses = InputProps && InputProps.classes;
    return /*#__PURE__*/_react.default.createElement(_TextField.default, (0, _extends2.default)({}, rest, {
      InputProps: (0, _extends2.default)({}, InputProps, {
        classes: (0, _extends2.default)({}, borderlessClasses, InputPropsClasses)
      }),
      ref: ref
    }));
  }

  if (variant === 'outlined') {
    console.warn("UI-Core: The TextField variant \"".concat(variant, "\" is not supported."));
    return /*#__PURE__*/_react.default.createElement(_TextField.default, (0, _extends2.default)({}, rest, {
      InputProps: InputProps,
      ref: ref
    }));
  }

  return /*#__PURE__*/_react.default.createElement(_TextField.default, (0, _extends2.default)({}, props, {
    ref: ref
  }));
});

TextField.defaultProps = {
  InputProps: undefined,
  select: undefined,
  variant: 'standard'
};
TextField.propTypes = process.env.NODE_ENV !== "production" ? {
  InputProps: _propTypes.default.shape({}),
  select: _propTypes.default.bool,
  variant: _propTypes.default.string
} : {};

var _default = (0, _utils.dataUICMaker)(TextField, 'TextField');

exports.default = _default;
