// @alteryx/ui/CardHeader/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _CardHeader = _interopRequireDefault(require("@material-ui/core/CardHeader"));

var _styles = require("../styles");

var _utils = require("../utils");

var useStyles = (0, _styles.makeStyles)(function (_ref) {
  var palette = _ref.palette;
  return {
    action: {
      marginTop: 1,
      paddingBottom: 1,
      borderBottom: "1px solid ".concat(palette.divider)
    }
  };
});

var CardHeader = /*#__PURE__*/_react.default.forwardRef(function (props, ref) {
  var classes = useStyles();
  var title = props.title;
  var titleIsTabs = title && title.type && title.type.muiName === 'Tabs';
  return /*#__PURE__*/_react.default.createElement(_CardHeader.default, (0, _extends2.default)({
    classes: titleIsTabs ? {
      action: classes.action
    } : undefined
  }, props, {
    ref: ref
  }));
});

CardHeader.propTypes = process.env.NODE_ENV !== "production" ? {
  title: _propTypes.default.node
} : {};
CardHeader.defaultProps = {
  title: null
};

var _default = (0, _utils.dataUICMaker)(CardHeader, 'CardHeader');

exports.default = _default;
