// @alteryx/ui/theme/props/prop-overrides.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireDefault(require("react"));

var _icons = require("@alteryx/icons");

var _Checkbox = _interopRequireDefault(require("../../internal/icons/Checkbox"));

var _CheckboxChecked = _interopRequireDefault(require("../../internal/icons/CheckboxChecked"));

var _CheckboxIndeterminate = _interopRequireDefault(require("../../internal/icons/CheckboxIndeterminate"));

var _Radio = _interopRequireDefault(require("../../internal/icons/Radio"));

var _RadioChecked = _interopRequireDefault(require("../../internal/icons/RadioChecked"));

var _Paper = _interopRequireDefault(require("../../Paper"));

var _ChevronDown = _interopRequireDefault(require("../../internal/icons/ChevronDown"));

var props = {
  MuiAccordion: {
    elevation: 0
  },
  MuiAccordionSummary: {
    expandIcon: /*#__PURE__*/_react.default.createElement(_icons.ChevronDown, null),
    IconButtonProps: {
      edge: false
    }
  },
  MuiAlert: {
    iconMapping: {
      error: /*#__PURE__*/_react.default.createElement(_icons.AlertCircle, {
        size: "small"
      }),
      info: /*#__PURE__*/_react.default.createElement(_icons.Info, {
        size: "small"
      }),
      warning: /*#__PURE__*/_react.default.createElement(_icons.AlertTriangle, {
        size: "small"
      }),
      success: /*#__PURE__*/_react.default.createElement(_icons.Check, {
        size: "small"
      })
    }
  },
  MuiAppBar: {
    elevation: 2
  },
  MuiAutocomplete: {
    popupIcon: /*#__PURE__*/_react.default.createElement(_icons.ChevronDown, {
      size: "small"
    }),
    closeIcon: /*#__PURE__*/_react.default.createElement(_icons.XSmall, {
      size: "small"
    }),
    ChipProps: {
      deleteIcon: /*#__PURE__*/_react.default.createElement(_icons.XSmall, null),
      size: 'small'
    },
    PaperComponent: /*#__PURE__*/_react.default.forwardRef(function (paperProps, ref) {
      return /*#__PURE__*/_react.default.createElement(_Paper.default, (0, _extends2.default)({
        elevation: 3,
        ref: ref
      }, paperProps));
    })
  },
  MuiButtonBase: {
    disableRipple: true
  },
  MuiButtonGroup: {
    disableRipple: true
  },
  MuiCard: {
    elevation: 1
  },
  MuiCardHeader: {
    titleTypographyProps: {
      variant: 'h2'
    }
  },
  MuiCheckbox: {
    icon: /*#__PURE__*/_react.default.createElement(_Checkbox.default, null),
    checkedIcon: /*#__PURE__*/_react.default.createElement(_CheckboxChecked.default, null),
    indeterminateIcon: /*#__PURE__*/_react.default.createElement(_CheckboxIndeterminate.default, null)
  },
  MuiCircularProgress: {
    size: 16
  },
  MuiDialog: {
    disableBackdropClick: true
  },
  MuiExpansionPanel: {
    elevation: 0
  },
  MuiExpansionPanelSummary: {
    expandIcon: /*#__PURE__*/_react.default.createElement(_icons.ChevronDown, null),
    IconButtonProps: {
      edge: false
    }
  },
  MuiFilledInput: {
    disableUnderline: true
  },
  MuiInput: {
    disableUnderline: true
  },
  MuiInputLabel: {
    shrink: true
  },
  MuiListItemText: {
    secondaryTypographyProps: {
      variant: 'subtitle2'
    }
  },
  MuiRadio: {
    color: 'primary',
    checkedIcon: /*#__PURE__*/_react.default.createElement(_RadioChecked.default, null),
    icon: /*#__PURE__*/_react.default.createElement(_Radio.default, null)
  },
  MuiSelect: {
    IconComponent: _ChevronDown.default,
    autoWidth: true,
    MenuProps: {
      getContentAnchorEl: null,
      anchorOrigin: {
        vertical: 'bottom',
        horizontal: 'center'
      },
      transformOrigin: {
        vertical: -4,
        horizontal: 'center'
      }
    }
  },
  MuiSpeedDialIcon: {
    icon: /*#__PURE__*/_react.default.createElement(_icons.Plus, null),
    openIcon: /*#__PURE__*/_react.default.createElement(_icons.X, null)
  },
  MuiStepper: {
    elevation: 2
  },
  MuiTablePagination: {
    rowsPerPageOptions: [50, 100, 250]
  }
};
var _default = props;
exports.default = _default;
