// @alteryx/ui/VirtualizedList/VirtualizedList.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof3 = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _typeof2 = _interopRequireDefault(require("@babel/runtime/helpers/typeof"));

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _reactWindow = require("react-window");

var _ui = require("@alteryx/ui");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof3(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var useStyles = (0, _ui.makeStyles)(function () {
  return {
    listbox: {
      '& ul': {
        marginBottom: 0
      }
    }
  };
});
var LIST_PADDING = 8;
var OuterElementContext = /*#__PURE__*/(0, _react.createContext)({});
var OuterElementType = /*#__PURE__*/(0, _react.forwardRef)(function (props, ref) {
  var outerProps = (0, _react.useContext)(OuterElementContext);
  var style = props.style;
  var newStyle = (0, _extends2.default)({}, style, {
    outline: 'none'
  });
  return /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({
    ref: ref,
    role: "listbox"
  }, props, outerProps, {
    style: newStyle,
    tabIndex: "0"
  }));
});
OuterElementType.defaultProps = {
  style: {}
};
OuterElementType.propTypes = process.env.NODE_ENV !== "production" ? {
  style: _propTypes.default.objectOf(_propTypes.default.any)
} : {};

var renderRow = function renderRow(_ref) {
  var data = _ref.data,
      index = _ref.index,
      style = _ref.style;

  if (! /*#__PURE__*/(0, _react.isValidElement)(data[index]) && (0, _typeof2.default)(data[index]) !== 'object') {
    return /*#__PURE__*/_react.default.createElement("span", {
      style: style
    }, "".concat(data[index]));
  }

  return /*#__PURE__*/(0, _react.cloneElement)(data[index], {
    style: (0, _extends2.default)({}, style, {
      top: style.top + LIST_PADDING
    })
  });
};

function useResetCache(data) {
  var ref = (0, _react.useRef)(null);
  (0, _react.useEffect)(function () {
    if (ref.current != null) {
      ref.current.resetAfterIndex(0, true);
    }
  }, [data]);
  return ref;
}

var VirtualizedList = /*#__PURE__*/(0, _react.forwardRef)(function (props, ref) {
  var children = props.children,
      itemDataProp = props.itemData,
      itemCountProp = props.itemCount,
      itemSizeProp = props.itemSize,
      renderRowProp = props.renderRow,
      prevIndex = props.prevIndex,
      direction = props.direction,
      other = (0, _objectWithoutProperties2.default)(props, ["children", "itemData", "itemCount", "itemSize", "renderRow", "prevIndex", "direction"]);
  var DEFAULT_ITEM_HEIGHT = 32;
  var DEFAULT_DISPLAY_COUNT = 8;
  var classes = useStyles();

  var itemData = itemDataProp || _react.default.Children.toArray(children);

  var itemCount = itemCountProp || itemData.length;
  var itemSize = itemSizeProp;

  var getHeight = function getHeight() {
    // 32 is our default item height(px), 8 is our default item display amount
    var maxHeight = itemSize() * itemCount > DEFAULT_ITEM_HEIGHT * DEFAULT_DISPLAY_COUNT ? DEFAULT_ITEM_HEIGHT * DEFAULT_DISPLAY_COUNT : itemSize() * itemCount || itemSize();
    return maxHeight;
  };

  var gridRef = useResetCache(itemCount);

  _react.default.useEffect(function () {
    var listbox = document.querySelector('[role="listbox"]');

    if (direction === 'ArrowUp' && prevIndex === 0) {
      // Using the listbox to scroll to the bottom so that a ref isn't removed.if
      if (listbox && listbox.scrollTo) {
        listbox.scrollTo(0, itemCount * itemSize());

        var setLastItemFocused = function setLastItemFocused() {
          var lastItem = document.querySelector("[data-option-index=\"".concat(itemCount - 1, "\"]"));

          if (lastItem !== null) {
            lastItem.setAttribute('data-focus', 'true');
          }
        };

        var lastFrame = requestAnimationFrame(setLastItemFocused);
        return function () {
          return cancelAnimationFrame(lastFrame);
        };
      }
    }

    if (direction === 'ArrowDown' && prevIndex === itemCount - 1) {
      // Using the listbox to scroll to the top so that a ref isn't removed.
      if (listbox && listbox.scrollTo) {
        listbox.scrollTo(0, 0);

        var setFirstItemFocus = function setFirstItemFocus() {
          var firstItem = document.querySelector("[data-option-index=\"".concat(0, "\"]"));

          if (firstItem !== null) {
            firstItem.setAttribute('data-focus', 'true');
          }
        };

        var firstFrame = requestAnimationFrame(setFirstItemFocus);
        return function () {
          return cancelAnimationFrame(firstFrame);
        };
      }
    } // eslint-disable-next-line react-hooks/exhaustive-deps

  }, [prevIndex, direction]);

  return /*#__PURE__*/_react.default.createElement("div", {
    className: classes.listbox,
    ref: ref
  }, /*#__PURE__*/_react.default.createElement(OuterElementContext.Provider, {
    value: other
  }, /*#__PURE__*/_react.default.createElement(_reactWindow.VariableSizeList, (0, _extends2.default)({
    height: getHeight(),
    innerElementType: "ul",
    itemCount: itemCount,
    itemData: itemData,
    itemSize: itemSize,
    outerElementType: OuterElementType,
    overscanCount: 5,
    ref: gridRef,
    width: "100%"
  }, other), renderRowProp || renderRow)));
});
VirtualizedList.defaultProps = {
  direction: '',
  itemCount: 0,
  itemData: undefined,
  itemSize: function itemSize() {
    return 32;
  },
  prevIndex: undefined,
  renderRow: renderRow
};
VirtualizedList.propTypes = process.env.NODE_ENV !== "production" ? {
  direction: _propTypes.default.string,
  itemCount: _propTypes.default.number,
  itemData: _propTypes.default.arrayOf(_propTypes.default.any),
  itemSize: _propTypes.default.func,
  prevIndex: _propTypes.default.number,
  renderRow: _propTypes.default.func
} : {};
var _default = VirtualizedList;
exports.default = _default;
